crates/hc_sandbox/src/zome_call.rs

```diff
@@ -0,0 +1,511 @@
+            CellInfo::Provisioned(info)
+                if info.cell_id.dna_hash() == zome_call.dna_hash.as_ref() =>
+            {
+                Some(info.cell_id.clone())
+            }
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    if cell_ids.is_empty() {
+        anyhow::bail!(
+            "No cell found for DNA hash [{:?}] in app {:?}",
+            zome_call.dna_hash,
+            info
+        );
+    }
+
+    holochain_util::pw::pw_set_piped(zome_call.piped);
+    if !zome_call.piped {
+        msg!("Enter passphrase to authorize zome calls: ");
+    }
+    let passphrase = holochain_util::pw::pw_get().context("Failed to get passphrase")?;
+
+    let (auth, key) = generate_signing_credentials(passphrase).await?;
+
+    let (nonce, expires_at) = holochain_nonce::fresh_nonce(Timestamp::now())
+        .map_err(|e| anyhow::anyhow!("Failed to generate nonce: {:?}", e))?;
+
+    let params = ZomeCallParams {
+        provenance: AgentPubKey::from_raw_32(key.verifying_key().as_bytes().to_vec()),
+        cell_id: cell_ids.first().unwrap().clone(),
+        zome_name: ZomeName::from(zome_call.zome_name.clone()),
+        fn_name: FunctionName(zome_call.function.clone()),
+        cap_secret: Some(auth.cap_secret),
+        payload: ExternIO::encode(serde_json::from_slice::<serde_json::Value>(
+            zome_call.payload.as_bytes(),
+        )?)?,
+        nonce,
+        expires_at,
+    };
+    let (payload, hash) = params.serialize_and_hash()?;
+
+    let sig = key.try_sign(&hash)?;
+
+    let response = client
+        .request(AppRequest::CallZome(Box::new(ZomeCallParamsSigned {
+            bytes: ExternIO::from(payload),
+            signature: Signature::try_from(sig.to_vec())?,
+        })))
+        .await?;
+
+    match response {
+        AppResponse::ZomeCalled(response) => {
+            let response: serde_json::Value = response.decode()?;
+            serde_json::to_writer(std::io::stdout(), &response)?;
+            println!(); // Add newline
+        }
+        other => anyhow::bail!("Unexpected response while calling zome: {:?}", other),
+    }
+
+    Ok(())
+}
+
+async fn generate_signing_credentials(
+    passphrase: BufRead,
+) -> anyhow::Result<(Auth, ed25519_dalek::SigningKey)> {
+    let auth = load_or_create_auth().await?;
+
+    let salt =
+        sodoken::BufReadSized::<{ sodoken::hash::argon2id::SALTBYTES }>::from(auth.salt.as_ref());
+
+    let hash = <sodoken::BufWriteSized<32>>::new_no_lock();
+    {
+        let read_guard = passphrase.read_lock();
+        sodoken::hash::argon2id::hash(
+            hash.clone(),
+            read_guard.to_vec(),
+            salt,
+            sodoken::hash::argon2id::OPSLIMIT_INTERACTIVE,
+            sodoken::hash::argon2id::MEMLIMIT_INTERACTIVE,
+        )
+        .await?;
+    }
+
+    Ok((
+        auth,
+        ed25519_dalek::SigningKey::from_bytes(hash.try_unwrap().unwrap().as_ref().try_into()?),
```

Comment:
epic conversion

Reply:
Yeah, I'm not 100% that I've converted this the most direct way, but it works :D